src/lib.rs

```diff
@@ -595,12 +642,87 @@ impl Cancelled {
 
 impl std::fmt::Display for Cancelled {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str("cancelled")
+        let why = match self {
+            Cancelled::PendingWrite => "pending write",
+            Cancelled::PropagatedPanic => "propagated panic",
+            Cancelled::UnexpectedCycle(_) => "unexpected cycle",
+        };
+        f.write_str("cancelled because of ")?;
+        f.write_str(why)
     }
 }
 
 impl std::error::Error for Cancelled {}
 
+/// Information about an "unexpected" cycle, meaning one where some of the
+/// participants lacked cycle recovery annotations.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Cycle {
+    participants: plumbing::CycleParticipants,
+}
+
+impl Cycle {
+    pub(crate) fn new(participants: plumbing::CycleParticipants) -> Self {
+        Self { participants }
+    }
+
+    /// Iterate over the [`DatabaseKeyIndex`] for each query participating
+    /// in the cycle. The start point of this iteration within the cycle
+    /// is arbitrary but deterministic, but the ordering is otherwise determined
+    /// by the execution.
+    pub fn participant_keys(&self) -> impl Iterator<Item = DatabaseKeyIndex> + '_ {
+        let min = self.participants.iter().min().unwrap();
+        let index = self.participants.iter().position(|p| p == min).unwrap();
+        self.participants[index..]
+            .iter()
+            .chain(self.participants[..index].iter())
+            .copied()
+    }
+
+    /// Returns a vector with the debug information for
+    /// all the participants in the cycle.
+    pub fn all_participants<DB: ?Sized + Database>(&self, db: &DB) -> Vec<String> {
+        self.participant_keys()
+            .map(|d| format!("{:?}", d.debug(db)))
+            .collect()
+    }
+
+    /// Returns a vector with the debug information for
+    /// those participants in the cycle that lacked recovery
+    /// information.
+    pub fn unexpected_participants<DB: ?Sized + Database>(&self, db: &DB) -> Vec<String> {
+        self.participant_keys()
+            .filter(|&d| db.cycle_recovery_strategy(d) == CycleRecoveryStrategy::Panic)
+            .map(|d| format!("{:?}", d.debug(db)))
+            .collect()
+    }
+
+    /// Returns a "debug" view onto this strict that can be used to print out information.
+    pub fn debug<'me, DB: ?Sized + Database>(&'me self, db: &'me DB) -> impl std::fmt::Debug + 'me {
+        struct UnexpectedCycleDebug<'me> {
+            c: &'me Cycle,
+            db: &'me dyn Database,
+        }
+
+        impl<'me> std::fmt::Debug for UnexpectedCycleDebug<'me> {
```

Comment:
I'd

![fmt](https://user-images.githubusercontent.com/1711539/139587973-3ec9bb0f-660e-415b-b08b-0e998ae80c43.gif)

here to not repeat std::fmt

Reply:
Hmm, I tend to feel the opposite: `std` is short enough that I get annoyed when people shorten it to `fmt` instead of `std::fmt`. That said, I think this aversion dates from the days when I didn't have rust-analyzer to manage the `use` statements for me =) 

No strong opinion, really.

Reply:
This particular code seems to have been removed in the final version anyway.